src/error.rs

```diff
@@ -1,56 +1,87 @@
-
 #[derive(Debug, PartialEq, Clone, Eq, Display)]
 pub enum Error {
-    #[display(fmt = "parse error")]
-    ParseError,
-    #[display(fmt = "unaligned page access")]
-    Unaligned,
-    #[display(fmt = "out of bound access")]
-    OutOfBound,
-    #[display(fmt = "max cycles exceeded")]
-    InvalidCycles,
-    #[display(fmt = "cycles overflow")]
+    #[display(fmt = "aot error: dynasm ret {}", "_0")]
+    AotDynasm(i32),
+    #[display(fmt = "aot error: section is empty")]
+    AotSectionIsEmpty,
+    #[display(fmt = "aot error: section overlaps with another")]
+    AotSectionOverlaps,
+    #[display(fmt = "aot error: limit reached maximum dummy sections")]
+    AotLimitReachedMaximumDummySections,
+    #[display(fmt = "aot error: limit reached maximum labels")]
+    AotLimitReachedMaximumLabels,
+    #[display(fmt = "aot error: limit reached maximum sections")]
+    AotLimitReachedMaximumSections,
+    #[display(fmt = "aot error: limit reached maximum temp register")]
+    AotLimitReachedMaximumTempRegisters,
+    #[display(fmt = "aot error: out of bound due to not start of basic block")]
+    AotOutOfBoundDueToNotStartOfBasicBlock,
+    #[display(fmt = "asm error: {}", "_0")]
+    Asm(u8),
+    #[display(fmt = "cycles error: max cycles exceeded")]
+    CyclesExceeded,
+    #[display(fmt = "cycles error: overflow")]
     CyclesOverflow,
+    #[display(fmt = "elf error: bits")]
+    ElfBits,
+    #[display(fmt = "elf error: {}", "_0")]
+    ElfParseError(String),
+    #[display(fmt = "elf error: segment is unreadable")]
+    ElfSegmentUnreadable,
+    #[display(fmt = "elf error: segment is writable and executable")]
+    ElfSegmentWritableAndExecutable,
+    #[display(fmt = "elf error: segment addr or size is wrong")]
+    ElfSegmentAddrOrSizeError,
+    // Unknown error type is for the debugging tool of CKB-VM, it should not be
+    // used in this project.
+    #[display(fmt = "external error: {}", "_0")]
+    External(String),
+    #[display(fmt = "invalid syscall {}", "_0")]
+    InvalidEcall(u64),
     #[display(
         fmt = "invalid instruction pc=0x{:x} instruction=0x{:x}",
         "pc",
         "instruction"
     )]
     InvalidInstruction { pc: u64, instruction: u32 },
-    #[display(fmt = "invalid syscall {}", "_0")]
-    InvalidEcall(u64),
-    #[display(fmt = "invalid elf")]
-    InvalidElfBits,
     #[display(fmt = "invalid operand {}", "_0")]
     InvalidOp(u16),
-    #[display(fmt = "I/O error: {:?}", "_0")]
-    IO(ErrorKind),
-    #[display(fmt = "dynasm error {}", "_0")]
-    Dynasm(i32),
-    #[display(fmt = "assembly error {}", "_0")]
-    Asm(u8),
-    #[display(fmt = "limit reached")] // FIXME: Distinguish which limit
-    LimitReached,
-    #[display(fmt = "invalid permission")] // FIXME: Distinguish which permission
-    InvalidPermission,
     #[display(fmt = "invalid version")]
     InvalidVersion,
+    #[display(fmt = "I/O error: {:?}", "_0")]
+    IO(std::io::ErrorKind),
```

Comment:
Will it be better if we preserve both the error kind, and a raw error message as string?

Reply:
Agree.